src/main/java/com/google/firebase/auth/Tenant.java

```diff
@@ -0,0 +1,202 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.auth;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.util.Key;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Contains metadata associated with a Firebase tenant.
+ *
+ * <p>Instances of this class are immutable and thread safe.
+ */
+public final class Tenant {
+
+  // Lazily initialized from 'resourceName'.
+  private String tenantId;
+
+  @Key("name")
+  private String resourceName;
+
+  @Key("displayName")
+  private String displayName;
+
+  @Key("allowPasswordSignup")
+  private boolean passwordSignInAllowed;
+
+  @Key("enableEmailLinkSignin")
+  private boolean emailLinkSignInEnabled;
+
+  public String getTenantId() {
+    if (tenantId == null) {
+      tenantId = resourceName.substring(resourceName.lastIndexOf("/") + 1);
+    }
+    return tenantId;
+  }
+
+  public String getDisplayName() {
+    return displayName;
+  }
+
+  public boolean isPasswordSignInAllowed() {
+    return passwordSignInAllowed;
+  }
+
+  public boolean isEmailLinkSignInEnabled() {
+    return emailLinkSignInEnabled;
+  }
+
+  /**
+   * Returns a new {@link UpdateRequest}, which can be used to update the attributes
+   * of this tenant.
+   *
+   * @return a non-null Tenant.UpdateRequest instance.
+   */
+  public UpdateRequest updateRequest() {
+    return new UpdateRequest(getTenantId());
+  }
+
+  /**
+   * A specification class for creating a new tenant.
+   *
+   * <p>Set the initial attributes of the new tenant by calling various setter methods available in
+   * this class. None of the attributes are required.
+   */
+  public static final class CreateRequest {
+
+    private final Map<String,Object> properties = new HashMap<>();
+
+    /**
+     * Creates a new {@link CreateRequest}, which can be used to create a new tenant.
+     *
+     * <p>The returned object should be passed to {@link TenantManager#createTenant(CreateRequest)}
+     * to register the tenant information persistently.
+     */
+    public CreateRequest() { }
+
+    /**
+     * Sets the display name for the new tenant.
+     *
+     * @param displayName a non-null, non-empty display name string.
+     */
+    public CreateRequest setDisplayName(String displayName) {
+      checkArgument(!Strings.isNullOrEmpty(displayName), "display name must not be null or empty");
+      properties.put("displayName", displayName);
+      return this;
+    }
+
+    /**
+     * Sets whether to allow email/password user authentication.
+     *
+     * @param passwordSignInAllowed a boolean indicating whether users can be authenticated using
+     *     an email and password, and false otherwise.
```

Comment:
Probably omit "and false otherwise" (Or "True if ... and false otherwise.")

here and below

Reply:
Omitted.